VideoWeb/VideoWeb/ClientApp/src/app/services/video-filter.service.ts

```diff
@@ -0,0 +1,185 @@
+import { Injectable } from '@angular/core';
+import { Camera } from '@mediapipe/camera_utils';
+import { Results, SelfieSegmentation } from '@mediapipe/selfie_segmentation';
+import { Observable, Subject } from 'rxjs';
+import { Logger } from './logging/logger-base';
+import { BackgroundFilter } from './models/background-filter';
+import { SessionStorage } from './session-storage';
+
+@Injectable({
+    providedIn: 'root'
+})
+export class VideoFilterService {
+    private readonly loggerPrefix = '[VideoFilterService] -';
+
+    private readonly preferredFilterCache: SessionStorage<BackgroundFilter>;
+    readonly PREFERRED_FILTER_KEY = 'vh.preferred.filter';
+
+    private _onFilterChanged = new Subject<BackgroundFilter | null>();
+    get onFilterChanged$(): Observable<BackgroundFilter | null> {
+        return this._onFilterChanged.asObservable();
+    }
+
+    videoElement: HTMLVideoElement;
+    canvasElement: HTMLCanvasElement;
+    canvasStream: MediaStream;
+
+    canvasCtx: CanvasRenderingContext2D;
+
+    filterOn: boolean;
+    selfieSegmentation: SelfieSegmentation;
+    activeFilter: BackgroundFilter;
+    imgs: Map<BackgroundFilter, HTMLImageElement> = new Map();
+
+    constructor(private logger: Logger) {
+        this.preferredFilterCache = new SessionStorage(this.PREFERRED_FILTER_KEY);
+
+        if (!this.preferredFilterCache.get()) {
+            this.filterOn = false;
+            this.activeFilter = null;
+        } else {
+            this.activeFilter = this.preferredFilterCache.get();
+            this.filterOn = true;
+        }
+
+        this.selfieSegmentation = new SelfieSegmentation({
+            locateFile: file => {
+                return `./scripts/selfie_segmentation/${file}`;
+            }
+        });
+        this.selfieSegmentation.setOptions({
+            modelSelection: 1,
+            selfieMode: true
+        });
+        this.selfieSegmentation.onResults(results => this.onSelfieSegmentationResults(results));
+    }
+
+    initFilterFromMediaStream(stream: MediaStream) {
+        if (this.videoElement && (this.videoElement?.srcObject as MediaStream)?.id !== stream.id) {
+            this.logger.debug(`${this.loggerPrefix} camera stream has changed`);
+            this.updateCameraStream(stream);
+        }
+        if (this.canvasStream) {
+            return;
+        }
+        this.logger.debug(`${this.loggerPrefix} initialising stream for filter`);
+        this.videoElement = document.createElement('video');
+        this.videoElement.srcObject = stream;
+
+        this.canvasElement = document.createElement('canvas');
+        this.canvasElement.width = 1280;
+        this.canvasElement.height = 720;
```

Comment:
Are these always the same? Could they change if someone is using phone? 

I see we also have the same width and height on lines 86/87. Do these need to be the same values? If so could we have them as consts?